geode-core/src/main/java/org/apache/geode/management/internal/configuration/domain/Configuration.java

```diff
@@ -168,80 +164,75 @@ public void toData(DataOutput out) throws IOException {
     DataSerializer.writeString(cacheXmlContent, out);
     DataSerializer.writeString(propertiesFileName, out);
     DataSerializer.writeProperties(gemfireProperties, out);
-    DataSerializer.writeHashSet((HashSet<?>) jarNames, out);
+    // before 1.12, we are wring a Hashset of jarnames to the output stream.
+    // as 1.12, write a null hashset to the stream, so that when we can still read the old
+    // configuration
+    DataSerializer.writeHashSet(null, out);
+    Version.getCurrentVersion().writeOrdinal(out, true);
+    DataSerializer.writeHashMap(deployments, out);
   }
 
   @Override
   public void fromData(DataInput in) throws IOException, ClassNotFoundException {
-    this.configName = DataSerializer.readString(in);
-    this.cacheXmlFileName = DataSerializer.readString(in);
-    this.cacheXmlContent = DataSerializer.readString(in);
-    this.propertiesFileName = DataSerializer.readString(in);
-    this.gemfireProperties = DataSerializer.readProperties(in);
-    this.jarNames = DataSerializer.readHashSet(in);
-  }
-
+    configName = DataSerializer.readString(in);
+    cacheXmlFileName = DataSerializer.readString(in);
+    cacheXmlContent = DataSerializer.readString(in);
+    propertiesFileName = DataSerializer.readString(in);
+    gemfireProperties = DataSerializer.readProperties(in);
+    HashSet<String> jarNames = DataSerializer.readHashSet(in);
+
+    Version version;
+    try {
+      // after 1.12, we would have version info in the input stream
+      version = Version.fromOrdinal(Version.readOrdinal(in));
+    } catch (Exception e) {
```

Comment:
Is there a gentler way to determine whether there's an ordinal? Some kind of "peek" or "is EOF" test to see if there's more data?

I looked briefly and couldn't find a way, but perhaps I overlooked something.

Reply:
DataInput does not have a way. So we would need to cast DataInput to some other class and it is hard to say if it would always be an instance of the other class.

Reply:
I think I found a way. Pre 1.12 the jarNames HashSet was never null. So we can use it to determine if we are Pre 1.12 (jarNames !=null) or 1.12 and later (jarNames == null). I will code up this fix.